File: apps/mobile-native/android/app/src/main/java/mx/nexara/mobile/nativeapp/data/integra/detection/IntegraSettingsRepository.kt
package mx.nexara.mobile.nativeapp.data.integra.detection

import android.content.Context
import mx.nexara.mobile.nativeapp.data.AuthRepository
import mx.nexara.mobile.nativeapp.data.api.ApiClient

/**
 * Ajustes de INTEGRA: sitios, sincronización y salud del abanico ACS.
 *
 * Las tres operaciones que cambian el mundo —[deleteSite], [runSync] y
 * [setModuleEnabled]— están separadas de las lecturas y llevan nombre explícito
 * para que en la pantalla no se llamen por accidente. Ninguna se ejecuta sin
 * que el operador haya visto antes lo que dice [deletionImpact] / [syncImpact]
 * / [moduleToggleImpact].
 */
class IntegraSettingsRepository(context: Context) {
    private val authRepo = AuthRepository(context)
    private val api: IntegraSettingsApi = ApiClient.authed(
        tokenProvider = { authRepo.token() },
        companyIdProvider = { authRepo.companyId() },
    ).create(IntegraSettingsApi::class.java)

    /* ── Lecturas ───────────────────────────────────────────────────── */

    suspend fun sites(): List<IntegraSite> = parseSites(IntegraJson.list(api.listSites()))

    suspend fun lastSync(siteId: Int): SyncRun? = parseSyncRun(IntegraJson.map(api.lastSync(siteId)))

    suspend fun acsFanout(siteId: Int? = null): List<AcsFanoutEntry> =
        IntegraJson.itemsOf(IntegraJson.map(api.acsFanoutStatus(siteId))).mapNotNull(::parseFanoutEntry)

    suspend fun capabilities(siteId: Int? = null): IntegraCapabilityCounts =
        parseCapabilityCounts(IntegraJson.map(api.capabilities(siteId)))

    /** Regiones del espejo: cuántas y cómo se llaman. */
    suspend fun regions(siteId: Int? = null): List<String> =
        IntegraJson.itemsOf(IntegraJson.map(api.regions(siteId))).mapNotNull { row ->
            (row["name"] as? String)?.takeIf { it.isNotBlank() }
                ?: (row["id"] as? String)?.takeIf { it.isNotBlank() }
        }

    /* ── Escrituras ─────────────────────────────────────────────────── */

    suspend fun createSite(draft: SiteDraft, isFirstSite: Boolean): IntegraSite? =
        parseSite(IntegraJson.map(api.createSite(siteCreateBody(draft, isFirstSite))))

    /**
     * Cambia un módulo del sitio. Se manda el mapa COMPLETO de overrides, no
     * solo la clave tocada: el servidor sustituye la columna entera y un envío
     * parcial apagaría en silencio todo lo que no viniera.
     */
    suspend fun setModuleEnabled(site: IntegraSite, moduleKey: String, enabled: Boolean): IntegraSite? {
        val merged = LinkedHashMap<String, Any?>()
        site.modulesOverride?.forEach { (k, v) -> merged[k] = v }
        merged[moduleKey] = enabled
        return parseSite(
            IntegraJson.map(api.updateSite(site.id, mapOf("modulesOverride" to merged))),
        )
    }

    suspend fun setActive(site: IntegraSite, active: Boolean): IntegraSite? =
        parseSite(IntegraJson.map(api.updateSite(site.id, mapOf("isActive" to active))))

    suspend fun makeDefault(site: IntegraSite): IntegraSite? =
        parseSite(IntegraJson.map(api.updateSite(site.id, mapOf("isDefault" to true))))

    suspend fun rename(site: IntegraSite, label: String): IntegraSite? =
        parseSite(IntegraJson.map(api.updateSite(site.id, mapOf("label" to label.trim()))))

    /** DESTRUCTIVO. Arrastra el inventario espejo del sitio. */
    suspend fun deleteSite(siteId: Int) {
        api.deleteSite(siteId)
    }

    /** Reconstruye el espejo desde los equipos. Tarda y carga el parque. */
    suspend fun runSync(siteId: Int): SyncRun? = parseSyncRun(IntegraJson.map(api.runSync(siteId)))
}

File: apps/mobile-native/ios/NexaraApp/Data/Integra/IntegraSettingsRepository.swift
import Foundation

/// Ajustes INTEGRA — sitios, sync y escrituras. Espejo de Android `IntegraSettingsRepository`.
enum IntegraSettingsRepository {
    static func sites() async throws -> [[String: Any]] {
        IntegraJSON.decodeList(try await IntegraHTTP.get("integra/sites"))
    }

    static func createSite(
        name: String,
        host: String,
        appKey: String,
        appSecret: String,
        label: String = "",
        provider: String = "ARTEMIS",
        isFirstSite: Bool
    ) async throws -> [String: Any] {
        var body: [String: Any] = [
            "name": name.trimmingCharacters(in: .whitespacesAndNewlines),
            "host": {
                var h = host.trimmingCharacters(in: .whitespacesAndNewlines)
                while h.hasSuffix("/") { h.removeLast() }
                return h
            }(),
            "appKey": appKey.trimmingCharacters(in: .whitespacesAndNewlines),
            "appSecret": appSecret.trimmingCharacters(in: .whitespacesAndNewlines),
            "provider": provider.uppercased(),
            "isDefault": isFirstSite,
        ]
        let lab = label.trimmingCharacters(in: .whitespacesAndNewlines)
        if !lab.isEmpty { body["label"] = lab }
        return IntegraJSON.decodeMap(try await IntegraHTTP.postMap("integra/sites", body: body))
    }

    static func updateSite(id: Int, body: [String: Any]) async throws -> [String: Any] {
        IntegraJSON.decodeMap(try await IntegraHTTP.patchMap("integra/sites/\(id)", body: body))
    }

    static func makeDefault(id: Int) async throws {
        _ = try await updateSite(id: id, body: ["isDefault": true])
        IntegraSiteScope.select(id)
    }

    static func setActive(id: Int, active: Bool) async throws {
        _ = try await updateSite(id: id, body: ["isActive": active])
    }

    static func rename(id: Int, label: String) async throws {
        _ = try await updateSite(id: id, body: ["label": label.trimmingCharacters(in: .whitespacesAndNewlines)])
    }

    /// Destructivo: arrastra el inventario espejo del sitio.
    static func deleteSite(id: Int) async throws {
        _ = try await IntegraHTTP.delete("integra/sites/\(id)")
        if IntegraSiteScope.current == id {
            IntegraSiteScope.select(nil)
        }
    }

    /// Reconstruye el espejo desde los equipos.
    static func runSync(siteId: Int) async throws -> [String: Any] {
        IntegraJSON.decodeMap(
            try await IntegraHTTP.postEmpty("integra/sync", query: ["siteId": "\(siteId)"])
        )
    }

    static func lastSync(siteId: Int) async throws -> [String: Any] {
        IntegraJSON.decodeMap(
            try await IntegraHTTP.get("integra/sync/last", query: ["siteId": "\(siteId)"])
        )
    }

    /// Salud del abanico ACS (`GET integra/acs-fanout/status`).
    static func acsFanout(siteId: Int? = nil) async throws -> [AcsFanoutEntry] {
        let root = IntegraJSON.decodeMap(
            try await IntegraHTTP.get("integra/acs-fanout/status", query: IntegraQuery.site(siteId))
        )
        return IntegraJSON.itemsOf(root).compactMap(AcsFanoutEntry.parse)
    }

    /// Conteos / módulos del espejo (`GET integra/capabilities`) — forma abierta.
    static func capabilities(siteId: Int? = nil) async throws -> IntegraCapabilityCounts {
        let root = IntegraJSON.decodeMap(
            try await IntegraHTTP.get("integra/capabilities", query: IntegraQuery.site(siteId))
        )
        return IntegraCapabilityCounts.parse(root)
    }

    /// Nombres de regiones del espejo (`GET integra/regions`).
    static func regions(siteId: Int? = nil) async throws -> [String] {
        let root = IntegraJSON.decodeMap(
            try await IntegraHTTP.get("integra/regions", query: IntegraQuery.site(siteId))
        )
        return IntegraJSON.itemsOf(root).compactMap { row in
            row.integraStr("name") ?? row.integraStr("id")
        }
    }

    static func siteDraftProblems(name: String, host: String, appKey: String, appSecret: String) -> [String] {
        var out: [String] = []
        if name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            out.append("El sitio necesita un nombre.")
        }
        let h = host.trimmingCharacters(in: .whitespacesAndNewlines)
        if h.isEmpty {
            out.append("Falta la dirección del servidor.")
        } else if !(h.hasPrefix("http://") || h.hasPrefix("https://")) {
            out.append("La dirección del servidor debe empezar por http:// o https://.")
        }
        if appKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            out.append("Falta la clave de aplicación (appKey).")
        }
        if appSecret.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            out.append("Falta el secreto de aplicación (appSecret).")
        }
        return out
    }
}
